File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.java
class Solution {
    public int characterReplacement(String s, int k) {
        HashMap<Character,Integer> count = new HashMap<>();
        int res = 0;
        int maxF = 0;
        int l = 0;

        for (int r=0;r<s.length();r++){
            count.put(s.charAt(r),1+count.getOrDefault(s.charAt(r),0));
            maxF = Math.max(maxF,count.get(s.charAt(r)));
            while(r-l+1 - maxF>k){
                count.put(s.charAt(l),count.getOrDefault(s.charAt(l),0)-1);
                l++;
            }
            res = Math.max(maxF,r-l+1);
        }return res;
        
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charSet = set()
        l=0
        count = 0
        for i in range(len(s)):
            while s[i] in charSet:
                charSet.remove(s[l])
                l+=1
            
            charSet.add(s[i])
            count = max(count,i-l+1)
        return count


